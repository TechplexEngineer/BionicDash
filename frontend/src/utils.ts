export type TreeNode = {
    name: string
    value: NTelement
    path: string
    children: TreeNode[]
    fullPath?: string
}

// network table keys
export type NTelement = {
    "key":           string,  //"/FMSInfo/.type",
    "value":         string,  //"\"FMSInfo\"",
    "type":          string,  //"string ",
    "id":            number,  //0,
    "sn":            number,  //1,
    "is_persistent": boolean, //false
    fullPath?: string //this seems like a hack... Maybe we should have a different type as the first arg to buildTree
};

// build a tree from the raw list of NTelements
export const buildTree = (elements: NTelement[], prefixToRemove?: string): TreeNode[] => {
    let tree: TreeNode[] = [];
    let level = {tree};

    if (prefixToRemove?.length) {
        elements.map(entry => {
            entry.fullPath = entry.key;
            entry.key = entry.key.substring(prefixToRemove.length);
            return entry;
        });
    }
    // console.log("-------Here----------")
    elements.forEach(path => {
        const a = path.key.split('/').reduce((r, name, idx, arr) => {
            // console.log("reducing:", path.key.split('/'), JSON.stringify(r), name)
            if (name === "") {
                return r
            }
            if (!r[name]) {
                r[name] = {tree: []};

                const obj = {
                    name,
                    value: {},
                    path:  arr.slice(0, idx+1).join("/"),
                    children: r[name].tree
                } as TreeNode;

                if (path.fullPath) {
                    const fullPathParts = path.fullPath.split("/")
                    const lastIdx = fullPathParts.lastIndexOf(name)
                    obj.fullPath = fullPathParts.slice(0, lastIdx+1).join("/").replace("/.metadata","")
                }

                // if we are processing the last path element
                if (path.key.endsWith(name)) { //same as idx == arr.length-1
                    // console.log(path, name)
                    obj.value = path
                }

                if (!Array.isArray(r.tree)) {
                    r.tree = [];
                }

                r.tree.push(obj);
            }
            return r[name];
        }, level)
        // console.log("ReducerResult",JSON.stringify(a))
    });
    // console.log(tree)
    return tree;
}
