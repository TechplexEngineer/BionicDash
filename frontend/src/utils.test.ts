import { assert, expect, test } from 'vitest'
import {buildTree, NTelement} from "./utils";



test('buildTree', () => {
  let elements = [
    {
      key: "Drivetrain/RF Speed/Size",
      fullPath: "/Shuffleboard/.metadata/Drivetrain/RF Speed/Size"
    },{
      key: "Drivetrain/RF Speed/Position",
      fullPath: "/Shuffleboard/.metadata/Drivetrain/RF uSpeed/Position"
    }
  ] as NTelement[]
  let tree = buildTree(elements)
  expect(tree).toMatchSnapshot()
})

test('buildTree without fullPath', () => {
  let elements = [
    {
      key: "Drivetrain/RF Speed/Size"
    },{
      key: "Drivetrain/RF Speed/Position"
    }
  ] as NTelement[]
  let tree = buildTree(elements)
  expect(tree).toMatchSnapshot()
})