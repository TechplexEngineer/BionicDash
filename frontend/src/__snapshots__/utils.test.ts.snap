// Vitest Snapshot v1

exports[`buildTree 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "fullPath": "/Shuffleboard/Drivetrain/RF Speed/Size",
            "name": "Size",
            "path": "Drivetrain/RF Speed/Size",
            "value": {
              "fullPath": "/Shuffleboard/.metadata/Drivetrain/RF Speed/Size",
              "key": "Drivetrain/RF Speed/Size",
            },
          },
          {
            "children": [],
            "fullPath": "/Shuffleboard/Drivetrain/RF uSpeed/Position",
            "name": "Position",
            "path": "Drivetrain/RF Speed/Position",
            "value": {
              "fullPath": "/Shuffleboard/.metadata/Drivetrain/RF uSpeed/Position",
              "key": "Drivetrain/RF Speed/Position",
            },
          },
        ],
        "fullPath": "/Shuffleboard/Drivetrain/RF Speed",
        "name": "RF Speed",
        "path": "Drivetrain/RF Speed",
        "value": {},
      },
    ],
    "fullPath": "/Shuffleboard/Drivetrain",
    "name": "Drivetrain",
    "path": "Drivetrain",
    "value": {},
  },
]
`;

exports[`buildTree without fullPath 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "name": "Size",
            "path": "Drivetrain/RF Speed/Size",
            "value": {
              "key": "Drivetrain/RF Speed/Size",
            },
          },
          {
            "children": [],
            "name": "Position",
            "path": "Drivetrain/RF Speed/Position",
            "value": {
              "key": "Drivetrain/RF Speed/Position",
            },
          },
        ],
        "name": "RF Speed",
        "path": "Drivetrain/RF Speed",
        "value": {},
      },
    ],
    "name": "Drivetrain",
    "path": "Drivetrain",
    "value": {},
  },
]
`;
